### PR TITLE
Fix metric name passed to external scaler in GetMetrics call. 

### LINKED ISSUE #2640: External scalers receive metric names with internal KEDA core prefix
### Discussed in https://github.com/kedacore/keda/discussions/2635

<div type='discussions-op-text'>

<sup>Originally posted by **krishna-kariya** February 14, 2022</sup>
Hi, 
I am testing HTTP-triggered Azure Functions on Azure Arc with Keda version 2.5 and 2.6.1. We have created our external HTTP scaler on Keda for autoscaling. 

I am getting the error below in the metrics API server 
`E0207 02:55:09.412436 1 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"No matching metrics found for s0-upstream_rq_total"}: No matching metrics found for s0-upstream_rq_total`.

The hpa shows a similar error: 
`invalid metrics (4 invalid out of 4), first error is: failed to get s0-upstream_rq_active external metric: unable to get external metric k8se-apps/s0-upstream_rq_active/&LabelSelector{MatchLabels:map[string]string{scaledobject.keda.sh/name: myapp,},MatchExpressions:[]LabelSelectorRequirement{},}: unable to fetch metrics from external metrics API: no matching metrics found for s0-upstream_rq_active` 

I see that from Keda v2.5 and later, the external metric names are prefixed with "sX-". Is this a new addition? 
We have created our external HTTP scaler. Do we need to update our external HTTP scaler with these new metric names? 

Any help would be appreciated. Thanks. </div>

## Tasklist

- [ ] Fix in KEDA core to pass correct metric name
- [ ] Provide e2e test(s)

End-users can mitigate this in their external scaler to remove the sX- prefix until we have shipped KEDA v2.6.2 or v2.7.0

### LINKED ISSUE #2697: Add e2e test for external scaler
Add e2e test for external scaler

### PR DESCRIPTION
**Pull request by v-shenoy**
_Provide a description of what has been changed_

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes #2640
Fixes #2697

### CODE REVIEW

File: pkg/scalers/external_scaler.go [hunk 1]
```diff
@@ -197,6 +197,8 @@ func (s *externalScaler) GetMetrics(ctx context.Context, metricName string, metr
 	}
 	defer done()
 
+	// Remove the sX- prefix as the external scaler shouldn't have to know about it
+	metricName = strings.SplitN(metricName, "-", 2)[1]
```
Comment (tomkerkhove) [inline pkg/scalers/external_scaler.go:201]: Thinking out loud - Might be good to add a unit test for this?

**Review by JorTurFer — COMMENTED** [review]
LGTM!
Only a test is needed :)

File: pkg/scalers/external_scaler.go (hunk 1 above)
Comment (zroubalik) [reply pkg/scalers/external_scaler.go:201]: +1
and might be actually better to define a function for this functionality next to exisiting one:
https://github.com/kedacore/keda/blob/d9172a74c43ceeafdc235142ddb817a1bc31f974/pkg/scalers/scaler.go#L109

Comment (v-shenoy) [reply pkg/scalers/external_scaler.go]: Done.

File: pkg/scalers/scaler.go [hunk 2]
```diff
@@ -110,6 +111,21 @@ func GenerateMetricNameWithIndex(scalerIndex int, metricName string) string {
 	return fmt.Sprintf("s%d-%s", scalerIndex, metricName)
 }
 
+// GenerateMetricNameWithoutIndex removes the index prefix from the metric name
+func GenerateMetricNameWithoutIndex(scalerIndex int, metricName string) (string, error) {
```
Comment (JorTurFer) [inline pkg/scalers/scaler.go:115]: I'd call this function something like `RemoveIndexFromMetricName`. It's more clear IMO.
WDYT?

**Review by JorTurFer — COMMENTED** [review]
LGTM,
Only small nit

File: pkg/scalers/scaler.go (hunk 2 above)
Comment (v-shenoy) [reply pkg/scalers/scaler.go:115]: Was my initial choice, then I decided to keep the naming similar. Will revert.

Comment (v-shenoy) [reply pkg/scalers/scaler.go]: Corrected.

**Review by zroubalik — APPROVED** [review]
LGTM, great job!

File: pkg/scalers/scaler_test.go [hunk 3]
```diff
@@ -85,3 +85,38 @@ func TestGetMetricTarget(t *testing.T) {
...
+		isError                              bool
+	}{
+		// Proper input
+		{scalerIndex: 0, metricName: "s0-metricName", expectedMetricNameWithoutIndexPrefix: "metricName", isError: false},
+		// Incorrect index prefix
+		{scalerIndex: 1, metricName: "s0-metricName", expectedMetricNameWithoutIndexPrefix: "", isError: true},
+		// Incorrect index prefix
+		{scalerIndex: 0, metricName: "0-metricName", expectedMetricNameWithoutIndexPrefix: "", isError: true},
+		// No index prefix
+		{scalerIndex: 0, metricName: "metricName", expectedMetricNameWithoutIndexPrefix: "", isError: true},
```
Comment (JorTurFer) [inline pkg/scalers/scaler_test.go:103]: Could you add a test case with index > 9? I mean, it's not common but it's a possibility that same ScaledObject have 10 or more scalers. I'm not sure if current implementation supports it

**Review by JorTurFer — APPROVED** [review]
LGTM!
Thanks a lot! ❤️

**Pull request merged**